char *mcast_html_string  = 
"<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\">\n"
"<html>\n"
"  <head>\n"
"    <meta charset=\"UTF-8\"/>\n"
"    <meta name=\"description\" content=\"Motif Cluster Alignment and Search Tool (MCAST) output.\"/>\n"
"    <title>MCAST</title>\n"
"    <script type=\"text/javascript\">\n"
"      var site_url = \"\";\n"
"\n"
"      var motifs = new Array();\n"
"      @motif_array@\n"
"\n"
"      var wrap = undefined;//size to display on one line\n"
"      var wrap_timer;\n"
"\n"
"      var seqmax = @max_match_length@;\n"
"\n"
"      var loadedSequences = new Array();\n"
"      //draging details\n"
"      var moving_seq;\n"
"      var moving_annobox;\n"
"      var moving_left;\n"
"      var moving_width;\n"
"      var moving_both;\n"
"\n"
"      //drag needles\n"
"      var dnl = null;\n"
"      var dnr = null;\n"
"      var drag_is_rc = undefined;\n"
"      //container\n"
"      var cont = null;\n"
"\n"
"      /*\n"
"       * $\n"
"       *\n"
"       * Shorthand function for getElementById\n"
"       */\n"
"      function $(el) {\n"
"        return document.getElementById(el);\n"
"      }\n"
"\n"
"      function mouseCoords(ev) {\n"
"        ev = ev || window.event;\n"
"        if(ev.pageX || ev.pageY){ \n"
"          return {x:ev.pageX, y:ev.pageY}; \n"
"        } \n"
"        return { \n"
"          x:ev.clientX + document.body.scrollLeft - document.body.clientLeft, \n"
"          y:ev.clientY + document.body.scrollTop  - document.body.clientTop \n"
"        };\n"
"      }\n"
"\n"
"      function setup() {\n"
"        rewrap();\n"
"        window.onresize = delayed_rewrap;\n"
"      }\n"
"\n"
"      function calculate_wrap() {\n"
"        var est_wrap = 0;\n"
"        var page_width;\n"
"        if (window.innerWidth) {\n"
"          page_width = window.innerWidth;\n"
"        } else if (document.body) {\n"
"          page_width = document.body.clientWidth;\n"
"        } else {\n"
"          page_width = null;\n"
"        }\n"
"        if (page_width) {\n"
"          var ruler_width = document.getElementById(\"ruler\").offsetWidth;\n"
"          est_wrap = Math.floor(page_width / ruler_width) * 5;\n"
"        }\n"
"        if (est_wrap > 20) {\n"
"          wrap = est_wrap - 20; \n"
"        } else {\n"
"          wrap = 100;\n"
"        }\n"
"      }\n"
"\n"
"      function rewrap() {\n"
"        var previous_wrap = wrap;\n"
"        calculate_wrap();\n"
"        if (previous_wrap != wrap) {\n"
"          for (var seqid in loadedSequences) {\n"
"            //exclude inherited properties and undefined properties\n"
"            if (!loadedSequences.hasOwnProperty(seqid) || loadedSequences[seqid] === undefined) continue;\n"
"          \n"
"            var sequence = loadedSequences[seqid];\n"
"            var annobox = document.getElementById(seqid + \"_annotation\");\n"
"            var leftPos = parseInt(document.getElementById(seqid + \"_dnl\").firstChild.firstChild.nodeValue);\n"
"            var rightPos = parseInt(document.getElementById(seqid + \"_dnr\").firstChild.firstChild.nodeValue);\n"
"            var is_rc = undefined;\n"
"            set_data(leftPos, rightPos - leftPos + 1, is_rc, sequence, annobox);\n"
"          }\n"
"        }\n"
"        if (wrap_timer) {\n"
"          clearTimeout(wrap_timer);\n"
"        }\n"
"        wrap_timer = setTimeout(\"rewrap()\", 5000);\n"
"      }\n"
"\n"
"      function delayed_rewrap() {\n"
"        if (wrap_timer) {\n"
"          clearTimeout(wrap_timer);\n"
"        }\n"
"        wrap_timer = setTimeout(\"rewrap()\", 1000);\n"
"      }\n"
"\n"
"      function showHidden(prefix) {\n"
"        document.getElementById(prefix + '_activator').style.display = 'none';\n"
"        document.getElementById(prefix + '_deactivator').style.display = 'block';\n"
"        document.getElementById(prefix + '_data').style.display = 'block';\n"
"      }\n"
"      function hideShown(prefix) {\n"
"        document.getElementById(prefix + '_activator').style.display = 'block';\n"
"        document.getElementById(prefix + '_deactivator').style.display = 'none';\n"
"        document.getElementById(prefix + '_data').style.display = 'none';\n"
"      }\n"
"\n"
"\n"
"      function show_more(seqid) {\n"
"        if (wrap == undefined) rewrap();\n"
"        \n"
"        var seq_row = document.getElementById(seqid + \"_blocks\");\n"
"        var info_row = document.getElementById(seqid + \"_info\");\n"
"        var tbl = document.getElementById(\"tbl_sequences\");\n"
"        if (info_row) {\n"
"          tbl.deleteRow(info_row.rowIndex);\n"
"          destroy_seq_handle(seqid, true);\n"
"          destroy_seq_handle(seqid, false);\n"
"          delete loadedSequences[seqid];\n"
"        } else {\n"
"          var sequence = new Sequence(seqid);\n"
"          var offset = sequence.segs[0].start;\n"
"          loadedSequences[seqid] = sequence;\n"
"          info_row = tbl.insertRow(seq_row.rowIndex + 1);\n"
"          info_row.id = seqid + \"_info\";\n"
"          var cell = info_row.insertCell(0);\n"
"          cell.colSpan = 9;\n"
"          cell.style.verticalAlign = \"top\";\n"
"          var width = Math.min(sequence.length, wrap); \n"
"          var start = 1;\n"
"          var is_neg_strand = false;\n"
"          //center on the first hit\n"
"          if (sequence.hits.length > 0) {\n"
"            var hit = sequence.hits[0];\n"
"            is_neg_strand = hit.is_rc;\n"
"            start = hit.pos - offset;\n"
"          }\n"
"          if (start + width >= sequence.length) start = sequence.length - width;\n"
"          create_info_section(cell, sequence, seqid, start+offset, width, is_neg_strand);\n"
"          var blockCont = document.getElementById(seqid + \"_block_container\");\n"
"          //create sequence view handles\n"
"          var dnl = create_seq_handle(blockCont, seqid, true);\n"
"          var dnr = create_seq_handle(blockCont, seqid, false);\n"
"          set_needles(offset, seqmax, dnl, start, dnr, start + width - 1);\n"
"        }\n"
"      }\n"
"\n"
"      function create_info_section(cell, sequence, seqid, start, width, is_neg_strand) {\n"
"        var box = document.createElement('div');\n"
"        box.className = \"box infobox\";\n"
"        //generic info\n"
"        box.appendChild(document.createTextNode(\"Change the portion of annotated sequence by \"));\n"
"        var bold_box = document.createElement('b');\n"
"        bold_box.appendChild(document.createTextNode(\"dragging the buttons\"));\n"
"        box.appendChild(bold_box);\n"
"        box.appendChild(document.createTextNode(\"; hold shift to drag them individually.\"));\n"
"        box.appendChild(document.createElement('br'));\n"
"        box.appendChild(document.createElement('br'));\n"
"        //sequence display\n"
"        var seqDispTitle = document.createElement('h5');\n"
"        seqDispTitle.appendChild(document.createTextNode(\"Annotated Sequence\"));\n"
"        box.appendChild(seqDispTitle);\n"
"        var show_rc_only = undefined;\n"
"        var annobox = document.createElement('div');\n"
"        annobox.id = seqid + \"_annotation\";\n"
"        set_data(start, width, show_rc_only, sequence, annobox);\n"
"        box.appendChild(annobox);\n"
"        cell.appendChild(box);\n"
"      }\n"
"\n"
"      function annobox_labels() {\n"
"        var seqDispLabel = document.createElement('div');\n"
"        seqDispLabel.className = \"sequence sequence_labels\";\n"
"        seqDispLabel.style.height = \"2.5em\";\n"
"        return seqDispLabel;\n"
"      }\n"
"\n"
"      function annobox_hits() {\n"
"        var seqDispHit = document.createElement('div');\n"
"        seqDispHit.className = \"sequence\";\n"
"        seqDispHit.style.height = \"1.5em\";\n"
"        return seqDispHit;\n"
"      }\n"
"\n"
"      function annobox_matches() {\n"
"        var seqDispMatch = document.createElement('div');\n"
"        seqDispMatch.className = \"sequence\";\n"
"        seqDispMatch.style.height = \"1.5em\";\n"
"        return seqDispMatch;\n"
"      }\n"
"\n"
"      function annobox_translations() {\n"
"        var seqDispXlate = document.createElement('div');\n"
"        seqDispXlate.className = \"sequence\";\n"
"        seqDispXlate.style.height = \"1.5em\";\n"
"        return seqDispXlate;\n"
"      }\n"
"\n"
"      function annobox_sequence() {\n"
"        var seqDispSeq = document.createElement('div');\n"
"        seqDispSeq.className = \"sequence\";\n"
"        seqDispSeq.style.height = \"1.5em\";\n"
"        return seqDispSeq; \n"
"      }\n"
"      function annobox_boundary() {\n"
"        var seqDispSeq = document.createElement('div');\n"
"        seqDispSeq.className = \"sequence\";\n"
"        seqDispSeq.style.height = \"1.5em\";\n"
"        return seqDispSeq; \n"
"      }\n"
"\n"
"      function create_seq_handle(container, seqid, isleft) {\n"
"        var vbar = document.createElement('div');\n"
"        vbar.id = seqid + \"_dn\" + (isleft ? \"l\" : \"r\");\n"
"        vbar.className = \"block_needle\" + (isleft ? \"\" : \" right\");\n"
"        var label = document.createElement('div');\n"
"        label.className = \"block_handle\" + (isleft ? \"\" : \" right\");\n"
"        label.appendChild(document.createTextNode(\"1\"));\n"
"        label.style.cursor = 'pointer';\n"
"        label.title = \"Drag to move the displayed range. Hold shift and drag to change \" + (isleft ? \"lower\" : \"upper\") + \" bound of the range.\";\n"
"        vbar.appendChild(label);\n"
"        container.appendChild(vbar);\n"
"        label.onmousedown = function(evt) {\n"
"          evt = evt || window.event;\n"
"          start_drag(seqid, isleft, !(evt.shiftKey));\n"
"        };\n"
"        return vbar;\n"
"      }\n"
"\n"
"      function destroy_seq_handle(seqid, isleft) {\n"
"        var vbar = document.getElementById(seqid + \"_dn\" + (isleft ? \"l\" : \"r\"));\n"
"        var label = vbar.firstChild;\n"
"        label.onmousedown = null;\n"
"        vbar.parentNode.removeChild(vbar);\n"
"      }\n"
"\n"
"      function start_drag(seqid, mouse_on_left, move_both) {\n"
"        //first record what we are moving\n"
"        moving_left = mouse_on_left;\n"
"        moving_both = move_both;\n"
"        moving_seq = loadedSequences[seqid];\n"
"        if (!moving_seq) {\n"
"          alert(\"Sequence not loaded?!\");\n"
"        }\n"
"        //get the container for movements to be judged against\n"
"        cont = document.getElementById(seqid + \"_block_container\");\n"
"        //get the div's that will be moved\n"
"        dnl = document.getElementById(seqid + \"_dnl\");\n"
"        dnr = document.getElementById(seqid + \"_dnr\");\n"
"        //get the div which has all the text containers that will be updated\n"
"        moving_annobox = document.getElementById(seqid + \"_annotation\");\n"
"        //calculate the space between handles\n"
"        moving_width = dnr.firstChild.firstChild.nodeValue - \n"
"            dnl.firstChild.firstChild.nodeValue;\n"
"        drag_is_rc = undefined;\n"
"        //setup the events for draging\n"
"        document.onmousemove = handle_drag;\n"
"        document.onmouseup = finish_drag;\n"
"      }\n"
"\n"
"\n"
"      function calculate_width(obj) {\n"
"        return obj.clientWidth - (obj.style.paddingLeft ? obj.style.paddingLeft : 0) \n"
"            - (obj.style.paddingRight ? obj.style.paddingRight : 0);\n"
"      }\n"
"\n"
"      function calculate_drag_pos(ev) {\n"
"        var mouse = mouseCoords(ev);\n"
"        var dragable_length = calculate_width(cont);\n"
"        //I believe that the offset parent is the body\n"
"        //otherwise I would need to make this recursive\n"
"        //maybe clientLeft would work, but the explanation of\n"
"        //it is hard to understand and it apparently doesn't work\n"
"        //in firefox 2.\n"
"        var diff = mouse.x - cont.offsetLeft;\n"
"        if (diff < 0) diff = 0;\n"
"        if (diff > dragable_length) diff = dragable_length;\n"
"        var pos = Math.round(diff / dragable_length * (seqmax));\n"
"        return pos;\n"
"      }\n"
"\n"
"      function handle_drag(ev) {\n"
"        var pos = calculate_drag_pos(ev);\n"
"        update_needles(pos, moving_seq.length, false);\n"
"      }\n"
"\n"
"      function finish_drag(ev) {\n"
"        document.onmousemove = null;\n"
"        document.onmouseup = null;\n"
"        var pos = calculate_drag_pos(ev);\n"
"        update_needles(pos, moving_seq.length, true);\n"
"      }\n"
"\n"
"      function update_needles(pos, seqlen, updateTxt) {\n"
"        var offset = moving_seq.segs[0].start;\n"
"        var leftPos = parseInt(dnl.firstChild.firstChild.nodeValue) - offset;\n"
"        var rightPos = parseInt(dnr.firstChild.firstChild.nodeValue) - offset;\n"
"        if (moving_both) {\n"
"          if (moving_left) {\n"
"            if ((pos + moving_width) >= seqlen) {\n"
"              pos = seqlen - moving_width - 1;\n"
"            }\n"
"            leftPos = pos;\n"
"            rightPos = pos + moving_width;\n"
"          } else {\n"
"            if (pos >= seqlen) {\n"
"              pos = seqlen - 1;\n"
"            } else if ((pos - moving_width) < 0) {\n"
"              pos = moving_width;\n"
"            }\n"
"            leftPos = pos - moving_width;\n"
"            rightPos = pos;\n"
"          }\n"
"        } else {\n"
"          if (moving_left) {\n"
"            if (pos > rightPos) {\n"
"              pos = rightPos;\n"
"            }\n"
"            leftPos = pos;\n"
"          } else {\n"
"            if (pos < leftPos) {\n"
"              pos = leftPos;\n"
"            } else if (pos >= seqlen) {\n"
"              pos = seqlen - 1;\n"
"            }\n"
"            rightPos = pos;\n"
"          }\n"
"        }\n"
"        set_needles(offset, seqmax, dnl, leftPos, dnr, rightPos);\n"
"        if (updateTxt) set_data(leftPos + offset, rightPos - leftPos + 1, drag_is_rc, moving_seq, moving_annobox);\n"
"      }\n"
"\n"
"      function set_needles(offset, max, left, leftPos, right, rightPos) {\n"
"        //the needles are between the sequence positions\n"
"        //the left needle is before and the right needle after\n"
"        //think of it as an inclusive range...\n"
"        left.style.left = \"\" + (leftPos / max * 100)+ \"%\";\n"
"        left.firstChild.firstChild.nodeValue=\"\" + (leftPos + offset);\n"
"        right.style.left = \"\" + ((rightPos+1) / max * 100)+ \"%\";\n"
"        right.firstChild.firstChild.nodeValue=\"\" + (rightPos + offset);\n"
"      }\n"
"\n"
"      function set_data(start, width, show_rc_only, data, annobox) {\n"
"        var child = annobox.firstChild;\n"
"        var line_width = Math.min(wrap, width);\n"
"        var num_per_wrap = 5\n"
"        var end = start + width;\n"
"        for (var i = start; i < end; i += line_width, line_width = Math.min(wrap, end - i)) {\n"
"          for (var j = 0; j < num_per_wrap; ++j) {\n"
"            if (child) {\n"
"              while (child.firstChild) child.removeChild(child.firstChild);\n"
"            } else {\n"
"              switch (j) {\n"
"              case 0:\n"
"                child = annobox_labels();\n"
"                break;\n"
"              case 1:\n"
"                child = annobox_hits();\n"
"                break;\n"
"              case 2:\n"
"                child = annobox_matches();\n"
"                break;\n"
"              case 3:\n"
"                child = annobox_sequence();\n"
"                break;\n"
"              case 4:\n"
"                child = annobox_boundary();\n"
"                break;\n"
"              }\n"
"              annobox.appendChild(child);\n"
"            }\n"
"            switch (j) {\n"
"            case 0:\n"
"              data.append_labels(child, i, line_width, show_rc_only);\n"
"              break;\n"
"            case 1:\n"
"              data.append_hits(child, i, line_width, show_rc_only);\n"
"              break;\n"
"            case 2:\n"
"              data.append_matches(child, i, line_width, show_rc_only);\n"
"              break;\n"
"            case 3:\n"
"              data.append_seq(child, i, line_width, show_rc_only);\n"
"              break;\n"
"            case 4:\n"
"              data.append_boundary(child, i, line_width, show_rc_only);\n"
"              break;\n"
"            }\n"
"            child = child.nextSibling;\n"
"          }\n"
"        }\n"
"        //clean up excess\n"
"        \n"
"        while (child) {\n"
"          var next = child.nextSibling;\n"
"          annobox.removeChild(child);\n"
"          child = next;\n"
"        }\n"
"      }\n"
"\n"
"      function append_coloured_nucleotide_sequence(container, sequence) {\n"
"        var len = sequence.length;\n"
"        for (var i = 0; i < len; i++) {\n"
"          var colour = \"black\";\n"
"          switch (sequence.charAt(i)) {\n"
"          case 'A':\n"
"            colour = \"red\";\n"
"            break;\n"
"          case 'C':\n"
"            colour = \"blue\";\n"
"            break;\n"
"          case 'G':\n"
"            colour = \"orange\";\n"
"            break;\n"
"          case 'T':\n"
"            colour = \"green\";\n"
"            break;\n"
"          }\n"
"          var letter = document.createElement('span');\n"
"          letter.style.color = colour;\n"
"          letter.appendChild(document.createTextNode(sequence.charAt(i)));\n"
"          container.appendChild(letter);\n"
"        }\n"
"      }\n"
"\n"
"      function append_coloured_peptide_sequence(container, sequence) {\n"
"        var len = sequence.length;\n"
"        for (var i = 0; i < len; i++) {\n"
"          var colour = \"black\";\n"
"          switch (sequence.charAt(i)) {\n"
"          case 'A':\n"
"          case 'C':\n"
"          case 'F':\n"
"          case 'I':\n"
"          case 'L':\n"
"          case 'V':\n"
"          case 'W':\n"
"          case 'M':\n"
"            colour = \"blue\";\n"
"            break;\n"
"          case 'N':\n"
"          case 'Q':\n"
"          case 'S':\n"
"          case 'T':\n"
"            colour = \"green\";\n"
"            break;\n"
"          case 'D':\n"
"          case 'E':\n"
"            colour = \"magenta\";\n"
"            break;\n"
"          case 'K':\n"
"          case 'R':\n"
"            colour = \"red\";\n"
"            break;\n"
"          case 'H':\n"
"            colour = \"pink\";\n"
"            break;\n"
"          case 'G':\n"
"            colour = \"orange\";\n"
"            break;\n"
"          case 'P':\n"
"            colour = \"yellow\";\n"
"            break;\n"
"          case 'Y':\n"
"            colour = \"turquoise\";\n"
"            break;\n"
"          }\n"
"          var letter = document.createElement('span');\n"
"          letter.style.color = colour;\n"
"          letter.appendChild(document.createTextNode(sequence.charAt(i)));\n"
"          container.appendChild(letter);\n"
"        }\n"
"      }\n"
"\n"
"      /*\n"
"       * Finds the index of an item in a sorted array (a) that equals a key (k)\n"
"       * when compared using a comparator (c). If an exact match can not be found\n"
"       * then returns -(index+1), where index is the location that the item would be inserted.\n"
"       */\n"
"      function bsearch(a, c, k) {\n"
"        var low = 0;\n"
"        var high = a.length;\n"
"        while (low < high) {\n"
"          var mid = low + Math.floor((high - low) /2);\n"
"          if (c(a[mid], k) < 0) {\n"
"            low = mid + 1;\n"
"          } else {\n"
"            high = mid;\n"
"          }\n"
"        }\n"
"        if ((low < a.length) && (c(a[low], k) == 0)) {\n"
"          return low;\n"
"        } else {\n"
"          return -(low+1);\n"
"        }\n"
"      }\n"
"\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // Start Motif Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      function Motif(name, type, best_f, best_rc) {\n"
"        this.name = name;\n"
"        this.is_nucleotide = (type === \"nucleotide\");\n"
"        this.best_f = best_f;\n"
"        this.best_rc = best_rc;\n"
"        this.length = best_f.length;\n"
"      }\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // End Motif Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // Start Seg Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      function Seg(start, segment) {\n"
"        this.start = parseInt(start);\n"
"        this.segment = segment;\n"
"      }\n"
"\n"
"      function compare_seg_to_pos(seg, pos) {\n"
"        return seg.start - pos;\n"
"      }\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // End Seg Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // Start Hit Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      function Hit(sequence_is_nucleotide, motif, pos, strand, pvalue, match) {\n"
"        //properties\n"
"        this.motif = motif;\n"
"        this.pos = parseInt(pos);\n"
"        this.width = 0;\n"
"        this.is_rc = (strand != \"+\");\n"
"        this.pvalue = pvalue;\n"
"        this.best = \"\";\n"
"        this.match = \"\";\n"
"        //functions\n"
"        this.find_overlap = Hit_find_overlap;\n"
"        //setup\n"
"        var seq;\n"
"        var m = match.replace(/ /g, \" \");\n"
"        if (this.is_rc) {\n"
"          seq = this.motif.best_rc;\n"
"        }\n"
"        else {\n"
"          seq = this.motif.best_f;\n"
"        }\n"
"        if (sequence_is_nucleotide == this.motif.is_nucleotide) {\n"
"          this.best = seq;\n"
"          this.match = m;\n"
"          this.width = motif.length;\n"
"        } else if (sequence_is_nucleotide) {\n"
"          this.width = motif.length * 3;\n"
"          for (var i = 0; i < motif.length; i++) {\n"
"            this.best += seq.charAt(i);\n"
"            this.best += \"..\";\n"
"            this.match += m.charAt(i);\n"
"            this.match += \"  \";\n"
"          }\n"
"        } else {\n"
"          throw \"UNSUPPORTED_CONVERSION\";\n"
"        }\n"
"      }\n"
"\n"
"      function Hit_find_overlap(sequence_is_nucleotide, start, width) {\n"
"        if (this.pos < start) {\n"
"          if ((this.pos + this.width) > start) {\n"
"            return {hit: this, start: start, length: Math.min(width, (this.pos + this.width - start))};\n"
"          } else {\n"
"            return null;\n"
"          }\n"
"        } else if (this.pos < (start + width)) {\n"
"          return {hit: this, start: this.pos, length: Math.min(this.width, (start + width - this.pos))};\n"
"        }\n"
"        return null;\n"
"      }\n"
"\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // End Hit Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // Start Overlapping Hit Iterator\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      function OLHIterator(sequence, start, width, is_rc) {\n"
"        //properties\n"
"        this.sequence = sequence;\n"
"        this.start = start;\n"
"        this.width = width;\n"
"        this.is_rc = is_rc;\n"
"        this.both = (is_rc === undefined);\n"
"        this.index = 0;\n"
"        this.nextO = null;\n"
"        //methods\n"
"        this.next = OLHIterator_next;\n"
"        this.has_next = OLHIterator_has_next;\n"
"        //setup\n"
"        // find the first hit which overlaps the range\n"
"        for (; this.index < this.sequence.hits.length; this.index++) {\n"
"          var hit = this.sequence.hits[this.index];\n"
"          if (!this.both && this.is_rc != hit.is_rc) continue;\n"
"          if ((this.nextO = hit.find_overlap(this.sequence.is_nucleotide, this.start, this.width)) != null) break;\n"
"        }\n"
"      }\n"
"      \n"
"      function OLHIterator_next() {\n"
"        if (this.nextO == null) throw \"NO_NEXT_ELEMENT\";\n"
"        var current = this.nextO;\n"
"        this.nextO = null;\n"
"        for (this.index = this.index + 1; this.index < this.sequence.hits.length; this.index++) {\n"
"          var hit = this.sequence.hits[this.index];\n"
"          if (!this.both && this.is_rc != hit.is_rc) continue;\n"
"          this.nextO = hit.find_overlap(this.sequence.is_nucleotide, this.start, this.width);\n"
"          break;\n"
"        }\n"
"        return current;\n"
"      }\n"
"      \n"
"      function OLHIterator_has_next() {\n"
"        return (this.nextO != null);\n"
"      }\n"
"\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // End Overlapping Hit Iterator\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // Start Sequence Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      function Sequence(seqid) {\n"
"        //properties\n"
"        this.length = parseInt(document.getElementById(seqid + \"_len\").value);\n"
"        this.cpvalue = document.getElementById(seqid + \"_combined_pvalue\").value;\n"
"        this.is_nucleotide = (document.getElementById(seqid + \"_type\").value === \"nucleotide\");\n"
"        this.segs = new Array(); //sorted list of segs\n"
"        this.hits = new Array(); //sorted list of hits\n"
"        //functions\n"
"        this.get_overlapping_hits = Sequence_get_overlapping_hits;\n"
"        this.append_seq = Sequence_append_seq;\n"
"        this.append_translation = Sequence_append_translation;\n"
"        this.append_hits = Sequence_append_hits;\n"
"        this.append_matches = Sequence_append_matches;\n"
"        this.append_labels = Sequence_append_labels;\n"
"        this.append_boundary = Sequence_append_boundary;\n"
"        //init\n"
"        //made this parser much more permissive as a new\n"
"        //version of libxml2 broke the translate call I was\n"
"        //using to remove whitespace characters\n"
"        var mysegs = document.getElementById(seqid + \"_segs\");\n"
"        var tokens = mysegs.value.split(/\\s+/);\n"
"        var offsetnum = \"\";\n"
"        var seqchunk = \"\";\n"
"        for (var i = 0; i < tokens.length; i++) {\n"
"          var token = tokens[i];\n"
"          if (token == \"\") continue;\n"
"          if (token.match(/\\d+/)) {\n"
"            if (offsetnum != \"\" && seqchunk != \"\") {\n"
"              this.segs.push(new Seg(offsetnum, seqchunk));\n"
"            }\n"
"            offsetnum = parseInt(token, 10);\n"
"            seqchunk = \"\";\n"
"            continue;\n"
"          }\n"
"          seqchunk += token;\n"
"        }\n"
"        if (offsetnum != \"\" && seqchunk != \"\") {\n"
"          this.segs.push(new Seg(offsetnum, seqchunk));\n"
"        }\n"
"        var myhits = document.getElementById(seqid + \"_hits\");\n"
"        lines = myhits.value.split(/\\n/);\n"
"        for (var i in lines) {\n"
"          //exclude inherited properties and undefined properties\n"
"          if (!lines.hasOwnProperty(i) || lines[i] === undefined) continue;\n"
"          \n"
"          var line = lines[i];\n"
"          var chunks = line.split(/\\t/);\n"
"          if (chunks.length != 5) continue;\n"
"          var pos = chunks[0];\n"
"          var motif = motifs[chunks[1]];\n"
"          var strand = chunks[2];\n"
"          var pvalue = chunks[3];\n"
"          var match = chunks[4];\n"
"          this.hits.push(new Hit(this.is_nucleotide, motif, pos, strand, pvalue, match));\n"
"        }\n"
"      }\n"
"\n"
"      function Sequence_get_overlapping_hits(start, width, is_rc) {\n"
"        return new OLHIterator(this, start, width, is_rc);\n"
"      }\n"
"\n"
"      function Sequence_append_seq(container, start, width, is_rc) {\n"
"        //make a sub container to put the sequence in\n"
"        var mycontainer = document.createElement('span');\n"
"        // find where a seg starting at start would be\n"
"        var i = bsearch(this.segs, compare_seg_to_pos, start);\n"
"        var seg;\n"
"        var sequence = \"\";\n"
"        if (i < -1) {\n"
"          //possible partial segment, need to handle first\n"
"          i = -(i + 1);\n"
"          seg = this.segs[i-1];\n"
"          if ((seg.start + seg.segment.length) > start) {\n"
"            var seg_offset = start - seg.start;\n"
"            var seg_width = Math.min(width, seg.segment.length - seg_offset);\n"
"            sequence += seg.segment.substring(seg_offset, seg_offset + seg_width);\n"
"          }\n"
"        } else if (i == -1) {\n"
"          //gap, with following segment, normal state at start of iteration\n"
"          i = 0;\n"
"        } \n"
"        for (;i < this.segs.length; i++) {\n"
"          seg = this.segs[i];\n"
"          var gap_width = Math.min(width - sequence.length, seg.start - (start + sequence.length));\n"
"          for (; gap_width > 0; gap_width--) sequence += '-';\n"
"          var seg_width = Math.min(width - sequence.length, seg.segment.length);\n"
"          if (seg_width == 0) break;\n"
"          sequence += seg.segment.substring(0, seg_width);\n"
"        }\n"
"        while (sequence.length < width) sequence += '-';\n"
"        // calculate which parts are aligned with hits and output them in colour\n"
"        var pos = start;\n"
"        var iter = this.get_overlapping_hits(start, width, is_rc);\n"
"        while (iter.has_next()) {\n"
"          var o = iter.next();\n"
"          var from, to;\n"
"          if (o.start > pos) {\n"
"            from = pos - start;\n"
"            to = o.start - start;\n"
"            var subseq = sequence.substring(from, to);\n"
"            var greytxt = document.createElement('span');\n"
"            greytxt.style.color = 'grey';\n"
"            greytxt.appendChild(document.createTextNode(subseq));\n"
"            mycontainer.appendChild(greytxt);\n"
"            pos = o.start;\n"
"          }\n"
"          from = pos - start;\n"
"          to = from + o.length;\n"
"          var motifseq = sequence.substring(from, to);\n"
"          if (this.is_nucleotide) {\n"
"            append_coloured_nucleotide_sequence(mycontainer, motifseq);\n"
"          } else {\n"
"            append_coloured_peptide_sequence(mycontainer, motifseq);\n"
"          }\n"
"          pos += o.length;\n"
"        }\n"
"        if (pos < (start + width)) {\n"
"          var greytxt = document.createElement('span');\n"
"          greytxt.style.color = 'grey';\n"
"          greytxt.appendChild(document.createTextNode(sequence.substring(pos - start)));\n"
"          mycontainer.appendChild(greytxt);\n"
"        }\n"
"        container.appendChild(mycontainer);\n"
"      }\n"
"\n"
"      function Sequence_append_translation(container, start, width, is_rc) {\n"
"        if (start > this.length) {\n"
"          alert(\"start: \" + start + \" length: \" + this.length);\n"
"          throw \"INDEX_OUT_OF_BOUNDS\";\n"
"        }\n"
"        if ((start + width - 1) > this.length) {\n"
"          alert(\"start: \" + start + \" width: \" + width + \" length: \" + this.length);\n"
"          throw \"RANGE_OUT_OF_BOUNDS\";\n"
"        }\n"
"        //make a sub container to put the sequence in\n"
"        var mycontainer = document.createElement('span');\n"
"        // calculate which parts are aligned with hits and output them in colour\n"
"        var pos = start;\n"
"        var iter = this.get_overlapping_hits(start, width, is_rc);\n"
"        while (iter.has_next()) {\n"
"          var o = iter.next();\n"
"          var space = \"\";\n"
"          var from, to;\n"
"          while (o.start > pos) {\n"
"            space += \"\\xA0\"\n"
"            ++pos;\n"
"          }\n"
"          if (space.length > 0) {\n"
"            var spacer = document.createElement('span');\n"
"            spacer.appendChild(document.createTextNode(space));\n"
"            mycontainer.appendChild(spacer);\n"
"          }\n"
"          from = o.start - o.hit.pos;\n"
"          to = from + o.length;\n"
"          if (o.hit.motif.is_nucleotide) {\n"
"            append_coloured_nucleotide_sequence(mycontainer, motifseq);\n"
"          } else {\n"
"            append_coloured_peptide_sequence(mycontainer, motifseq);\n"
"          }\n"
"          pos += o.length;\n"
"        }\n"
"\n"
"        container.appendChild(mycontainer);\n"
"      }\n"
"\n"
"      function Sequence_append_matches(container, start, width, is_rc) {\n"
"        //make a sub container to put the sequence in\n"
"        var mycontainer = document.createElement('span');\n"
"        var pos = start;\n"
"        var text = \"\";\n"
"        var iter = this.get_overlapping_hits(start, width, is_rc);\n"
"        while (iter.has_next()) {\n"
"          var o = iter.next();\n"
"          var space = \"\";\n"
"          var from, to;\n"
"          while (o.start > pos) {\n"
"            text += \"\\xA0\"\n"
"            ++pos;\n"
"          }\n"
"          from = o.start - o.hit.pos;\n"
"          to = from + o.length;\n"
"          var motifseq = o.hit.match.substring(from, to);\n"
"          text += motifseq\n"
"          pos += o.length;\n"
"        }\n"
"        mycontainer.appendChild(document.createTextNode(text));\n"
"        container.appendChild(mycontainer);\n"
"      }\n"
"\n"
"      function Sequence_append_hits(container, start, width, is_rc) {\n"
"        //make a sub container to put the sequence in\n"
"        var mycontainer = document.createElement('span');\n"
"        // calculate which parts are aligned with hits and output them in colour\n"
"        var pos = start;\n"
"        var iter = this.get_overlapping_hits(start, width, is_rc);\n"
"        while (iter.has_next()) {\n"
"          var o = iter.next();\n"
"          var space = \"\";\n"
"          var from, to;\n"
"          while (o.start > pos) {\n"
"            space += \"\\xA0\"\n"
"            ++pos;\n"
"          }\n"
"          if (space.length > 0) {\n"
"            var spacer = document.createElement('span');\n"
"            spacer.appendChild(document.createTextNode(space));\n"
"            mycontainer.appendChild(spacer);\n"
"          }\n"
"          from = o.start - o.hit.pos;\n"
"          to = from + o.length;\n"
"          var motifseq = o.hit.best.substring(from, to);\n"
"          if (o.hit.motif.is_nucleotide) {\n"
"            append_coloured_nucleotide_sequence(mycontainer, motifseq);\n"
"          } else {\n"
"            append_coloured_peptide_sequence(mycontainer, motifseq);\n"
"          }\n"
"          pos += o.length;\n"
"        }\n"
"\n"
"        container.appendChild(mycontainer);\n"
"      }\n"
"\n"
"      function Sequence_append_labels(container, start, width, is_rc) {\n"
"        //create a table (ye gods...)\n"
"        var oTable = document.createElement(\"table\");\n"
"        var oRow = oTable.insertRow(oTable.rows.length);\n"
"\n"
"        // calculate where to put the labels\n"
"        var pos = start;\n"
"        var cellindex = 0;\n"
"        var iter = this.get_overlapping_hits(start, width, is_rc);\n"
"        while (iter.has_next()) {\n"
"          var o = iter.next();\n"
"          var motif_center = Math.floor(o.start + o.length / 2);\n"
"          var spacer = \"\";\n"
"          while (pos < motif_center) {\n"
"            spacer += \"\\xA0\"\n"
"            pos++;\n"
"          }\n"
"          var cell1 = oRow.insertCell(cellindex++);\n"
"          cell1.appendChild(document.createTextNode(spacer));\n"
"          //add one for the div\n"
"          pos++;\n"
"          var cell2 = oRow.insertCell(cellindex++);\n"
"          //create the div that holds the label divs\n"
"          var div_container = document.createElement('div');\n"
"          div_container.className = \"label_container\";\n"
"          cell2.appendChild(div_container);\n"
"          //create the top label\n"
"          var top_label = \"\";\n"
"          top_label += \"(\";\n"
"          if (o.hit.is_rc) {\n"
"            top_label += \"-\";\n"
"          } else {\n"
"            top_label += \"+\";\n"
"          }\n"
"          top_label += \") \";\n"
"          if (o.hit.motif.name.match(/^\\d+$/)) {\n"
"            top_label += \"Motif \" + o.hit.motif.name;\n"
"          } else {\n"
"            top_label += o.hit.motif.name;\n"
"          }\n"
"          var div_top_label = document.createElement('div');\n"
"          div_top_label.className = \"sequence_label sequence_label_top\";\n"
"          div_top_label.style.width = \"\" + top_label.length + \"em\";\n"
"          div_top_label.style.left = \"-\" + (top_label.length / 2) + \"em\";\n"
"          div_top_label.appendChild(document.createTextNode(top_label));\n"
"          //create the bottom label\n"
"          var bottom_label = o.hit.pvalue;\n"
"          var div_bottom_label = document.createElement('div');\n"
"          div_bottom_label.className = \"sequence_label sequence_label_bottom\";\n"
"          div_bottom_label.style.width = \"\" + bottom_label.length + \"em\";\n"
"          div_bottom_label.style.left = \"-\" + (bottom_label.length / 2) + \"em\";\n"
"          div_bottom_label.appendChild(document.createTextNode(bottom_label));\n"
"          //add the label divs to the positioned container\n"
"          div_container.appendChild(div_top_label);\n"
"          div_container.appendChild(div_bottom_label);\n"
"        }\n"
"        container.appendChild(oTable);\n"
"\n"
"      }\n"
"\n"
"      function Sequence_append_boundary(container, start, width, is_rc) {\n"
"        //make a sub container to put the sequence in\n"
"        var mycontainer = document.createElement('span');\n"
"        var pos = start;\n"
"        var text = \"\";\n"
"        var iter = this.get_overlapping_hits(start, width, is_rc);\n"
"        while (iter.has_next()) {\n"
"          var o = iter.next();\n"
"          var end;\n"
"          while (o.start > pos) {\n"
"            text += \"\\xA0\"\n"
"            ++pos;\n"
"          }\n"
"          if (o.start == o.hit.pos) {\n"
"            text += \"\\\\\";\n"
"            ++pos;\n"
"          }\n"
"          end = o.start + o.length - 1;\n"
"          while (end > pos) {\n"
"            text += \"_\";\n"
"            ++pos;\n"
"          }\n"
"          if (end == (o.hit.pos + o.hit.width -1)) {\n"
"            text += \"/\"\n"
"          } else {\n"
"            text += \"_\"\n"
"          }\n"
"          ++pos;\n"
"        }\n"
"        mycontainer.appendChild(document.createTextNode(text));\n"
"        container.appendChild(mycontainer);\n"
"      }\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"      // End Sequence Object\n"
"      //+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n"
"    </script>\n"
"    <style type=\"text/css\">\n"
"        body { background-color:white; font-size: 12px; font-family: Verdana, Arial, Helvetica, sans-serif;}\n"
"\n"
"        div.help {\n"
"          display: inline-block;\n"
"          margin: 0px;\n"
"          padding: 0px;\n"
"          width: 12px;\n"
"          height: 13px;\n"
"          cursor: pointer;\n"
"          background-image: url(data:image/gif;base64,R0lGODlhDAANAIABANR0AP///yH5BAEAAAEALAAAAAAMAA0AAAIdhI8Xy22MIFgv1DttrrJ7mlGNNo4c+aFg6SQuUAAAOw==);\n"
"        }\n"
"\n"
"        div.help:hover {\n"
"          background-image: url(data:image/gif;base64,R0lGODlhDAANAKEAANR0AP///9R0ANR0ACH+EUNyZWF0ZWQgd2l0aCBHSU1QACH5BAEAAAIALAAAAAAMAA0AAAIdDGynCe3PgoxONntvwqz2/z2K2ImjR0KhmSIZUgAAOw==);\n"
"        }\n"
"        \n"
"        p.spaced { line-height: 1.8em;}\n"
"        \n"
"        span.citation { font-family: \"Book Antiqua\", \"Palatino Linotype\", serif; color: #004a4d;}\n"
"\n"
"        p.pad { padding-left: 30px; padding-top: 5px; padding-bottom: 10px;}\n"
"\n"
"        td.jump { font-size: 13px; color: #ffffff; background-color: #00666a;\n"
"          font-family: Georgia, \"Times New Roman\", Times, serif;}\n"
"\n"
"        a.jump { margin: 15px 0 0; font-style: normal; font-variant: small-caps;\n"
"          font-weight: bolder; font-family: Georgia, \"Times New Roman\", Times, serif;}\n"
"\n"
"        h2.mainh {font-size: 1.5em; font-style: normal; margin: 15px 0 0;\n"
"          font-variant: small-caps; font-family: Georgia, \"Times New Roman\", Times, serif;}\n"
"\n"
"        h2.line {border-bottom: 1px solid #CCCCCC; font-size: 1.5em; font-style: normal;\n"
"          margin: 15px 0 0; padding-bottom: 3px; font-variant: small-caps;\n"
"          font-family: Georgia, \"Times New Roman\", Times, serif;}\n"
"\n"
"        h4 {border-bottom: 1px solid #CCCCCC; font-size: 1.2em; font-style: normal;\n"
"          margin: 10px 0 0; padding-bottom: 3px; font-family: Georgia, \"Times New Roman\", Times, serif;}\n"
"\n"
"        h5 {margin: 0px}\n"
"\n"
"        a.help { font-size: 9px; font-style: normal; text-transform: uppercase;\n"
"          font-family: Georgia, \"Times New Roman\", Times, serif;}\n"
"\n"
"        div.pad { padding-left: 30px; padding-top: 5px; padding-bottom: 10px;}\n"
"        \n"
"        div.pad1 { margin: 10px 5px;}\n"
"\n"
"        div.pad2 { margin: 25px 5px 5px;}\n"
"        h2.pad2 { padding: 25px 5px 5px;}\n"
"\n"
"        div.pad3 { padding: 5px 0px 10px 30px;}\n"
"\n"
"        div.box { border: 2px solid #CCCCCC; padding:10px; overflow: hidden;}\n"
"\n"
"        div.bar { border-left: 7px solid #00666a; padding:5px; margin-top:25px; }\n"
"\n"
"        div.subsection {margin:25px 0px;}\n"
"\n"
"        img {border:0px none;}\n"
"\n"
"        th.majorth {text-align:left;}\n"
"        th.minorth {font-weight:normal; text-align:left; width:8em; padding: 3px 0px;}\n"
"        th.actionth {font-weight:normal; text-align:left;}\n"
"\n"
"        .block_td {height:25px;}\n"
"        .block_container {position:relative; width:98%; height:25px; padding:0px; margin: 0px 0px 0px 1em;}\n"
"        .block_motif {position:absolute; z-index:3; height:12px; top:0px; text-align:center; vertical-align:middle; background-color:cyan;}\n"
"        .block_rule {position:absolute; z-index:2; width:100%; height:1px; top:12px; left:0px; background-color:gray;}\n"
"        .block_plus_sym {position:absolute; z-index:4; line-height:12px; top:0px; left:-1em;}\n"
"        .block_minus_sym {position:absolute; z-index:4; line-height:12px; top:13px; left:-1em;}\n"
"\n"
"        .tic_major {position:absolute; border-left:2px solid blue; height:0.5em; top:0em;}\n"
"        .tic_minor {position:absolute; border-left:1px solid blue; height:0.2em; top:0em;}\n"
"        .tic_label {position:absolute; top:0.5em;  height: 1em; text-align:center; vertical-align:middle}\n"
"\n"
"        .explain h5 {font-size:1em; margin-left: 1em;}\n"
"\n"
"        div.doc {margin-left: 2em; margin-bottom: 3em;}\n"
"        \n"
"        th.trainingset {\n"
"          border-bottom: thin dashed black; \n"
"          font-weight:normal; \n"
"          padding:0px 10px;\n"
"        }\n"
"        .dnaseq {\n"
"          font-weight: bold; \n"
"          font-size: large; \n"
"          font-family: 'Courier New', Courier, monospace;\n"
"        }\n"
"        .dna_A {\n"
"          color: rgb(204,0,0);\n"
"        }\n"
"        .dna_C {\n"
"          color: rgb(0,0,204);\n"
"        }\n"
"        .dna_G {\n"
"          color: rgb(255,179,0);\n"
"        }\n"
"        .dna_T {\n"
"          color: rgb(0,128,0);\n"
"        }\n"
"        div.pop_content {\n"
"          position:absolute;\n"
"          z-index:50;\n"
"          width:300px;\n"
"          padding: 5px;\n"
"          background: #E4ECEC;\n"
"          font-size: 12px;\n"
"          font-family: Arial;\n"
"          border-style: double;\n"
"          border-width: 3px;\n"
"          border-color: #AA2244;\n"
"          display:none;\n"
"        }\n"
"        div.pop_content > div.pop_close {\n"
"          float:right;\n"
"          bottom:0px;\n"
"        }\n"
"\n"
"        /*****************************************************************************\n"
"         * Program logo styling\n"
"         ****************************************************************************/\n"
"        div.prog_logo {\n"
"          border-bottom: 0.25em solid #0f5f60;\n"
"          height: 4.5em;\n"
"          width: 24em;\n"
"          display:inline-block;\n"
"        }\n"
"        div.prog_logo img {\n"
"          float:left;\n"
"          width: 4em;\n"
"          border-style: none;\n"
"          margin-right: 0.2em;\n"
"        }\n"
"        div.prog_logo h1, div.prog_logo h1:hover, div.prog_logo h1:active, div.prog_logo h1:visited {\n"
"          margin:0;\n"
"          padding:0;\n"
"          font-family: Arial, Helvetica,  sans-serif;\n"
"          font-size: 3.2em;\n"
"          line-height: 1em;\n"
"          vertical-align: top;\n"
"          display: block;\n"
"          color: #026666;\n"
"          letter-spacing: -0.06em;\n"
"          text-shadow: 0.04em 0.06em 0.05em #666;\n"
"        }\n"
"        div.prog_logo h2, div.prog_logo h2:hover, div.prog_logo h2:active, div.prog_logo h2:visited {\n"
"          display: block;\n"
"          margin:0;\n"
"          padding:0;\n"
"          font-family: Helvetica, sans-serif;\n"
"          font-size: 0.9em;\n"
"          line-height: 1em;\n"
"          letter-spacing: -0.06em;\n"
"          color: black;\n"
"        }\n"
"\n"
"        div.big.prog_logo {\n"
"          font-size: 18px;\n"
"        }\n"
"\n"
"      .more_arrow {\n"
"        font-family:Arial,Serif;\n"
"        font-size: larger;\n"
"        font-weight: bold;\n"
"        text-decoration:none; \n"
"      }\n"
"      div.infobox {\n"
"        background-color:#ddddff;\n"
"        margin-top: 50px;\n"
"        margin-bottom: 1em;\n"
"      }\n"
"      .sequence {font-family:Monospace;}\n"
"      .sequence_labels {position:relative; width:100%; height:2.5em; \n"
"          padding:0px; margin:0px;}\n"
"      .sequence_label {font-family:Serif; position:absolute; z-index:2; \n"
"          height:1em; text-align:center; vertical-align:middle;}\n"
"      .sequence_label_top {top:0px;}\n"
"      .sequence_label_bottom {top:1.25em;}\n"
"      .inlineTitle {display:inline;}\n"
"      .block_needle {position:absolute; z-index:4; height:30px; width:1px; \n"
"          top:-2px; background-color:gray;}\n"
"      .block_handle {position:absolute; z-index:5; height:1.1em; width:7em; \n"
"          top:30px; left:-4em; text-align:center; vertical-align:middle;\n"
"          background-color: LightGrey; border:3px outset grey; \n"
"          -moz-user-select: none; -khtml-user-select: none; user-select: none;}\n"
"      .block_needle.right {height: 50px;}\n"
"      .block_handle.right {top:50px;}\n"
"      .label_container {position:relative; width:100%; height:25px; padding:0px; margin: 0px;}\n"
"\n"
"      table.padded-table td { padding:0px 10px; }\n"
"      table.padded-table th { padding:0px 5px; }\n"
"      td.tnum {text-align:right;}\n"
"      tr.highlight {background:#aaffaa;}\n"
"      td.dim {color: gray;}\n"
"      span.dim {color: gray;}\n"
"      .col_seq { text-align: left; width: @longest_seq_name@}\n"
"      .col_start, .col_stop, .col_score, .col_pv, .col_ev, .col_qv { text-align:right; }\n"
"      .col_start {width: 8em}\n"
"      .col_stop {width: 8em}\n"
"      .col_score {width: 5em}\n"
"      .col_pv {width: 5em}\n"
"      .col_ev {width: 5em}\n"
"      .col_qv {width: 5em}\n"
"      .col_more {width: 1em}\n"
"      .col_extra {width: 1em}\n"
"      .col_bd {width: 100%}\n"
"    </style>\n"
"  </head>\n"
"  <body onload=\"javascript:setup()\" >\n"
"    <!-- Top -->\n"
"    <a name=\"top\"/>\n"
"    <div class=\"pad1\">\n"
"      <div class=\"prog_logo big\">\n"
"        <img src=\"\" alt=\"MCAST Logo\" id=\"title_icon\">\n"
"        <script>$(\"title_icon\").src = site_url + \"/doc/images/mcast_icon.png\";</script>\n"
"        <h1>MCAST</h1>\n"
"        <h2>Motif Cluster Alignment and Search Tool</h2>\n"
"      </div>\n"
"      <p class=\"spaced\">\n"
"        For further information on how to interpret these results or to get a \n"
"        copy of the MEME software please access \n"
"        <a href=\"http://meme.nbcr.net/\">http://meme.nbcr.net</a>. \n"
"      </p>\n"
"      <p>\n"
"        If you use MCAST in your research please cite the following paper:<br /><span class=\"citation\">\n"
"          Timothy Bailey and William Stafford Noble, \n"
"          &quot;Searching for statistically significant regulatory modules&quot;, \n"
"          <i>Bioinformatics (Proceedings of the European Conference on Computational Biology)</i>,\n"
"          19(Suppl. 2):ii16-ii25, 2003.\n"
"          <a href=\"http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/suppl_2/ii16\">[full text]</a></span></p>\n"
"    </div>\n"
"    <a name=\"navigation\"/></a><div class=\"pad2\">\n"
"      <a class=\"jump\" href=\"#inputs\">Inputs</a>\n"
"        |  \n"
"      <a class=\"jump\" href=\"#results\">Search Results</a>\n"
"        |  \n"
"      <a class=\"jump\" href=\"#program\">Program information</a>\n"
"        |  \n"
"      <a class=\"jump\" href=\"#doc\">Documentation</a>\n"
"    </div>    \n"
"    <a name=\"inputs\"/>\n"
"    <table width=\"100%\" border=\"0\" cellspacing=\"1\" cellpadding=\"4\" bgcolor=\"#FFFFFF\">\n"
"      <tr>\n"
"        <td>\n"
"          <h2 class=\"mainh\">Inputs</h2>\n"
"        </td>\n"
"        <td align=\"right\" valign=\"bottom\">\n"
"          <a href=\"#top\">Top</a>\n"
"        </td>\n"
"      </tr>\n"
"    </table>\n"
"    <div class=\"box\">\n"
"      <!-- databases -->\n"
"      <a name=\"databases\"/>\n"
"      <h4>Sequence Database <a href=\"#databases_doc\" class=\"help\"><div class=\"help\"/></a></h4>\n"
"      <div class=\"pad\">\n"
"        <p>\n"
"          The sequence database &quot;@sequence_filename@&quot;\n"
"          was supplied to MCAST.\n"
"          </p>\n"
"          <table class=\"padded-table\" border=\"0\" >\n"
"            <col />\n"
"            <col />\n"
"            <thead>\n"
"              <tr>\n"
"                <th>Sequence Count</th>\n"
"                <th>Residue Count</th>\n"
"              </tr>\n"
"            </thead>\n"
"            <tbody>\n"
"                <tr>\n"
"                  <td class=\"tnum\">@num_sequences@</td>\n"
"                  <td class=\"tnum\">@num_residues@</td>\n"
"                </tr>\n"
"            </tbody>\n"
"            <tfoot>\n"
"            </tfoot>\n"
"          </table>\n"
"        </div>\n"
"        <!-- motifs -->\n"
"        <a name=\"motifs\" />\n"
"        <h4>Motifs <a href=\"#motifs_doc\" class=\"help\"><div class=\"help\"/></a></h4>\n"
"        <div class=\"pad\">\n"
"          <p>\n"
"            The following motifs were supplied to MCAST from &quot;@pattern_filename@&quot;\n"
"          </p>\n"
"          <table class=\"padded-table\" border=\"0\">\n"
"            <thead>\n"
"            <tr>\n"
"            <th>Motif</th>\n"
"            <th>Width</th>\n"
"            <th>Best possible match</th>\n"
"            <th>Best possible RC match</th>\n"
"            </tr>\n"
"            </thead>\n"
"            <tbody>\n"
"              @motif_list@\n"
"            </tbody>\n"
"          </table>\n"
"        </div>\n"
"      </div>\n"
"      <!-- Results -->\n"
"      <a name=\"results\"/>\n"
"      <table width=\"100%\" border=\"0\" cellspacing=\"1\" cellpadding=\"4\" bgcolor=\"#FFFFFF\">\n"
"        <tr>\n"
"          <td>\n"
"            <h2 class=\"mainh\">Search Results</h2>\n"
"          </td>\n"
"          <td align=\"right\" valign=\"bottom\">\n"
"            <a href=\"#top\">Top</a>\n"
"          </td>\n"
"        </tr>\n"
"      </table>\n"
"      <div class=\"box\">\n"
"        <h4>Top Scoring Matches <a href=\"#sequences_doc\" class=\"help\"><div class=\"help\"/></a></h4>\n"
"        <div class=\"pad\">\n"
"          @results@\n"
"        </div>\n"
"      </div>\n"
"      <!-- Program -->\n"
"      <a name=\"program\"/>\n"
"      <div class=\"bar\">\n"
"        <div style=\"text-align:right;\"><a href=\"#top\">Top</a></div>\n"
"        <div class=\"subsection\">\n"
"          <a name=\"version\"/>\n"
"          <h5>MCAST version</h5>\n"
"          @version@ (Release date: @release_date@)\n"
"        </div>\n"
"        <div class=\"subsection\">\n"
"          <a name=\"command\" />\n"
"          <h5>Command line summary</h5>\n"
"          <textarea rows=\"1\" style=\"width:100%;\" readonly=\"readonly\">\n"
"            @command_line@\n"
"          </textarea>\n"
"          <br />\n"
"          Background letter frequencies (from @bg_source@):\n"
"          <br/>\n"
"          <div style=\"margin-left:25px;\">\n"
"          @background_freqs@\n"
"          </div>\n"
"          <br />\n"
"          @computation_time@\n"
"          <br />\n"
"        </div>      \n"
"      <!-- Documentation -->\n"
"      <span class=\"explain\">\n"
"      <a name=\"doc\"/>\n"
"      <table width=\"100%\" border=\"0\" cellspacing=\"1\" cellpadding=\"4\" bgcolor=\"#FFFFFF\">\n"
"        <tr>\n"
"          <td>\n"
"            <h2 class=\"mainh\">Explanation of MCAST Results</h2>\n"
"          </td>\n"
"          <td align=\"right\" valign=\"bottom\">\n"
"            <a href=\"#top\">Top</a>\n"
"          </td>\n"
"        </tr>\n"
"      </table>\n"
"        <div class=\"box\">\n"
"          <h4>The MCAST results consist of</h4>\n"
"          <ul>\n"
"            <li>The <a href=\"#input_motifs_doc\"><b>inputs</b></a> to MCAST including:\n"
"              <ol>\n"
"                <li>The <a href=\"#databases_doc\"><b>sequence database</b></a> showing the sequence \n"
"                  and residue counts. [<a href=\"#databases\">View</a>]</li>\n"
"                <li>The <a href=\"#motifs_doc\"><b>motifs</b></a> showing the name, width,\n"
"                and best scoring match [<a href=\"#motifs\">View</a>]</li> \n"
"              </ol>\n"
"            </li>\n"
"            <li>The <a href=\"#sequences_doc\"><b>search results</b></a> showing top scoring sequences with \n"
"              tiling of all of the motifs matches shown for each of the sequences. [<a href=\"#results\">View</a>]\n"
"            </li>\n"
"            <li>The <b>program</b> details including:\n"
"              <ol>\n"
"                <li>The <b>version</b> of MCAST and the date it was released. [<a href=\"#version\">View</a>]</li>\n"
"                <li>The <b>command line summary</b> detailing the parameters with which you ran MCAST. [<a href=\"#command\">View</a>]</li>\n"
"              </ol>\n"
"            </li>\n"
"            <li>This <b>explanation</b> of how to interpret MCAST results.</li>\n"
"          </ul>\n"
"          <a name=\"input_motifs_doc\"/>\n"
"          <h4>Inputs</h4>\n"
"          <p>MCAST received the following inputs.</p>\n"
"          <a name=\"databases_doc\"/>\n"
"          <h5>Sequence Databases</h5>\n"
"          <div class=\"doc\">\n"
"            <p>This table summarises the sequence databases specified to MCAST.</p>\n"
"            <dl>\n"
"              <dt>Database</dt>\n"
"              <dd>The name of the database file.</dd>\n"
"              <dt>Sequence Count</dt>\n"
"              <dd>The number of sequences in the database.</dd>\n"
"              <dt>Residue Count</dt>\n"
"              <dd>The number of residues in the database.</dd>\n"
"            </dl>\n"
"          </div>\n"
"          <a name=\"motifs_doc\"/>\n"
"          <h5>Motifs</h5>\n"
"          <div class=\"doc\">\n"
"            <p>Summary of the motifs specified to MCAST.</p>\n"
"            <dl>\n"
"              <dt>Name</dt>\n"
"              <dd>The name of the motif. If the motif has been removed or removal is recommended to avoid highly similar motifs \n"
"                then it will be displayed in red text.</dd>\n"
"              <dt>Width</dt>\n"
"              <dd>The width of the motif. No gaps are allowed in motifs supplied to MCAST as it only works for motifs of a fixed width.</dd>\n"
"              <dt>Best possible match</dt>\n"
"              <dd>The sequence that would achieve the best possible match score and its reverse complement for nucleotide motifs.</dd>\n"
"            </dl>\n"
"          </div>\n"
"          <a name=\"sequences_doc\"/>\n"
"          <h4>Search Results</h4>\n"
"          <p>MCAST provides the following motif search results.</p>\n"
"          <h5>Top Scoring Sequences</h5>\n"
"          <div class=\"doc\">\n"
"            <p>\n"
"              This table summarises the top scoring sequences.\n"
"              The sequences are sorted by the <a href=\"#evalue_doc\">Sequence \n"
"              <i>q</i>-value</a> from most to least significant.\n"
"            </p>\n"
"            <dl>\n"
"              <dt>Sequence</dt>\n"
"              <dd>The name of the sequence. This maybe be linked to search a sequence database for the sequence name.</dd>\n"
"              <dt>Start</dt>\n"
"              <dd>\n"
"                The position of the start of the match in 1-based coordinates, starting at the\n"
"                first position in the sequence.\n"
"              </dd>\n"
"              <dt>Stop</dt>\n"
"              <dd>\n"
"                The position of the end of the match in 1-based coordinates, starting at the\n"
"                first position in the sequence.\n"
"              </dd>\n"
"              <dt>Score</dt>\n"
"              <dd>\n"
"                The MCAST score for the match.\n"
"              </dd>\n"
"              <dt><i>p</i>-value</dt>\n"
"              <dd>\n"
"                The <a href=\"#pvalue_doc\"><i>p</i>-value</a> of the match score.\n"
"              </dd>\n"
"              <dt><i>E</i>-value</dt>\n"
"              <dd>\n"
"                The <a href=\"#evalue_doc\"><i>E</i>-value</a> of the match score.\n"
"              </dd>\n"
"              <dt><i>q</i>-value</dt>\n"
"              <dd>\n"
"                The <a href=\"#qvalue_doc\"><i>q</i>-value</a> of the match score.\n"
"              </dd>\n"
"              <dt>&more;</dt>\n"
"              <dd>\n"
"                Click on this to show <a href=\"#additional_seq_doc\">additional information</a> about the sequence such as a \n"
"                description, combined p-value and the annotated sequence.\n"
"              </dd>\n"
"              <dt>Block Diagram</dt>\n"
"              <dd>\n"
"                The block diagram shows the best non-overlapping tiling of motif matches on the sequence.\n"
"                <ul style=\"padding-left: 1em; margin-top:5px;\">\n"
"                  <li>The length of the line shows the length of a sequence relative to all the other sequences.</li>\n"
"                  <li>A block is shown where the <a href=\"#pos_pvalue_doc\" >positional <i>p</i>-value</a>\n"
"                    of a motif is less (more significant) than the significance threshold which is 0.0001 by default.</li>\n"
"                  <li>If a significant motif match (as specified above) overlaps other significant motif matches then \n"
"                    it is only displayed as a block if its <a href=\"#pos_pvalue_doc\" >positional <i>p</i>-value</a> \n"
"                    is less (more significant) then the product of the <a href=\"#pos_pvalue_doc\" >positional \n"
"                    <i>p</i>-values</a> of the significant matches that it overlaps.</li>\n"
"                  <li>The position of a block shows where a motif has matched the sequence.</li>\n"
"                  <li>The width of a block shows the width of the motif relative to the length of the sequence.</li>\n"
"                  <li>The colour and border of a block identifies the matching motif as in the legend.</li>\n"
"                  <li>The height of a block gives an indication of the significance of the match as \n"
"                    taller blocks are more significant. The height is calculated to be proportional \n"
"                    to the negative logarithm of the <a href=\"#pos_pvalue_doc\" >positional <i>p</i>-value</a>, \n"
"                    truncated at the height for a <i>p</i>-value of 1e-10.</li>\n"
"                    <li>Hovering the mouse cursor over the block causes the display of the motif name \n"
"                      and other details in the hovering text.</li>\n"
"                  <li>DNA only; blocks displayed above the line are a match on the given DNA, whereas blocks \n"
"                    displayed below the line are matches to the reverse-complement of the given DNA.</li>\n"
"                  <li>DNA only; when strands are scored separately then blocks may overlap on opposing strands.</li>\n"
"                </ul>\n"
"              </dd>\n"
"            </dl>\n"
"          </div>\n"
"          <a name=\"additional_seq_doc\"/>\n"
"        </div>\n"
"      </span>\n"
"    <!-- This data is used by the javascript to create the detailed views -->\n"
"    <form>\n"
"      @mcast_match_data@\n"
"    </form>\n"
"    <span class=\"sequence\" id=\"ruler\" style=\"visibility:hidden; white-space:nowrap;\">ACGTN</span>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"    <br/>\n"
"  </body>\n"
"</html>\n"
"";

/******************
void fimo_print_motif_array();
void fimo_print_max_match_length();
void fimo_print_longest_seq_name();
void fimo_print_sequence_filename();
void fimo_print_num_sequences();
void fimo_print_num_residues();
void fimo_print_pattern_filename();
void fimo_print_motif_list();
void fimo_print_results();
void fimo_print_version();
void fimo_print_release_date();
void fimo_print_command_line();
void fimo_print_bg_source();
void fimo_print_background_freqs();
void fimo_print_computation_time();
void fimo_print_mcast_match_data();
******************/